src/address.rs

```diff
@@ -145,125 +148,136 @@ impl<'s> AddressParser<'s> {
     }
 
     #[stable]
-    pub fn parse_address_list(&mut self) -> Vec<Address> {
+    pub fn parse_address_list(&mut self) -> ParsingResult<Vec<Address>> {
         let mut result = Vec::new();
+        let mut expected_separator: char;
 
         while !self.p.eof() {
             self.p.push_position();
 
-            let mut entry = self.parse_group();
-
-            if entry.is_none() {
-                // If we failed to parse as group, try again as mailbox
-                self.p.pop_position();
-
-                entry = match self.parse_mailbox() {
-                    Some(mailbox) => Some(Address::Mailbox(mailbox)),
-                    None => None,
+            match self.parse_group() {
+                Ok(x) => {
+                    // Is a group
+                    result.push(x);
+                    expected_separator = ';';
+                },
+                Err(e) => {
+                    // If we failed to parse as group, try again as mailbox
+                    self.p.pop_position();
+                    result.push(Address::Mailbox(match self.parse_mailbox() {
+                        Ok(x) => x,
+                        Err(e2) => return Err(ParsingError::new(
```

Comment:
I'd consider partial results to be a separate feature (which I doubt is worth the implementation effort), and the use of results over options to be a strict improvement. The original reason I implemented this was because some trait from rust itself made the switch from options to results.